numpy/core/src/umath/string_ufuncs.cpp

```diff
@@ -0,0 +1,444 @@
+#include <Python.h>
+
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+#define _UMATHMODULE
+
+#include "numpy/ndarraytypes.h"
+
+#include "numpyos.h"
+#include "dispatching.h"
+#include "dtypemeta.h"
+#include "common_dtype.h"
+#include "convert_datatype.h"
+
+
+template <typename character>
+static NPY_INLINE int
+character_cmp(character a, character b)
+{
```

Comment:
Yeah, although I am not sure in case of invalid unicode characters (they are probably unsigned, but that would be out of range).  Happy to change if think it makes it much clearer?  (I don't have an opinion either way.)

Reply:
yeah, maybe something other than an in is better as `typename std::intermediate_type<npy_ucs4, npy_byte>::type`

Reply:
I honestly still don't really see that this will lead to easier to read code :), and I assume the compilers should optimize it just fine?

Reply:
Not quite, and that's why I'm trying to convince you :-)

https://godbolt.org/z/GMWPKE4EK

Reply:
Yeah, but that is not what we are comparing?  We are comparing these:
https://godbolt.org/z/f5qh9a3b7

And there `func1` actually ends up shorter in some cases (depends on the operand)?